.github/workflows/publish.yml

```
@@ -215,7 +215,7 @@ jobs:
           name: SARIF ${{ matrix.arch }}
           path: ${{ steps.scan.outputs.sarif }}
       - name: Upload SARIF file
-        if: always() && steps.scan.outputs.sarif != ''
+        if: '!cancelled() && steps.scan.outputs.sarif != '''''
```

Can we use `( … )` instead of the `' … '` everywhere as this
- firstly is common to use when condition start with a `!`
- secondly avoids the "is this a string literal or an expression" question
- thirdly avoids the need to escape the inner single quotes here and below

```suggestion
        if: ( !cancelled() && steps.scan.outputs.sarif != '' )
```

---

> firstly is common to use when condition start with a !

Not so common, I have never seen that used anywhere before, but should work just the same, yes, or using the explicit expression syntax `${{ !cancelled() }}`, whatever you like best in your project.

> secondly avoids the "is this a string literal or an expression" question

This question does not really exist and should never be asked usually.
It **is** a string either way, and it **is** interpreted as expression, that's just how it works.

---

Exactly, or rather both.
The file format is YAML for which this is (or must be) a `String` (and if it starts with a `!` it is instead interpreted as a YAML tag).
GitHub Actions then interprets at supported place (unfortunately not everywhere) `${{ ... }}` within those strings as expressions, evaluating their content and filling in the result, and in even more special places like `if`, the whole value is interpreted as an expression even without the explicit expression syntax.